marshal.go

```diff
@@ -1400,13 +1400,19 @@ func marshalList(info TypeInfo, value interface{}) ([]byte, error) {
 	return nil, marshalErrorf("can not marshal %T into %s", value, info)
 }
 
-func readCollectionSize(info CollectionType, data []byte) (size, read int) {
+func readCollectionSize(info CollectionType, data []byte) (size, read int, err error) {
 	if info.proto > protoVersion2 {
+		if len(data) < 4 {
+			return 0, 0, unmarshalErrorf("unmarshal list: unexpected eof")
+		}
 		size = int(data[0])<<24 | int(data[1])<<16 | int(data[2])<<8 | int(data[3])
 		read = 4
 	} else {
 		size = int(data[0])<<8 | int(data[1])
 		read = 2
+		if len(data) < 2 {
```

Comment:
Could you please move that if lines up? Otherwise it won't trigger, since the bound check from the compiler will trigger first.

A test for that should confirm this.

Reply:
Thanks for noticing! I've fixed the commit and added a separate one with a test in case you want to include it as well.